lib/pos_staking_rewards.go

```diff
@@ -1,5 +1,271 @@
+
+	// For case 1, we distribute the rewards by adding them to the staker's staked amount.
+	if stakeEntry != nil && stakeEntry.RewardMethod == StakingRewardMethodRestake {
+		stakeEntry.StakeAmountNanos.Add(stakeEntry.StakeAmountNanos, uint256.NewInt().SetUint64(rewardAmount))
+		bav._setStakeEntryMappings(stakeEntry)
+
+		return nil
+	}
+
+	// For cases 2 and 3, the staker does not want their rewards restaked. The staker is still
+	// eligible to receive rewards because their stake was used to secure the network. So we pay out
+	// the rewards directly to the staker's wallet.
+
+	stakerPublicKey := bav.GetPublicKeyForPKID(stakerPKID)
+	if _, err = bav._addBalance(rewardAmount, stakerPublicKey); err != nil {
+		return errors.Wrapf(err, "distributeStakingReward: problem adding rewards to staker's DESO balance: ")
+	}
+
 	return nil
 }
+
+func (bav *UtxoView) distributeValidatorCommission(validatorPKID *PKID, commissionAmount uint64) error {
+	// Here, we treat the validator's commission identically to staking rewards. We view commissions as another source of staking
+	// rewards that validators receive at the end of each epoch. And these commissions are eligible to be restaked if the validator
+	// desires. To determine whether to re-stake commissions or pay out the commissions to the validator's wallet, we rely on the
+	// validators own StakeEntry where they have staked to themselves, and the RewardMethod flag on the entry. The logic works as follows:
+	// - If the validator has staked to themselves, and they have reward restaking enabled, then their commissions are restaked.
+	// - If the validator has not staked to themselves, or they have reward restaking disabled, then their commissions are paid out
+	//   to their wallet.
+	//
+	// This approach has a few advantages:
+	// 1. It gives validators an easy opt-in feature to restake their commissions. This is useful for validators that want to maximize
+	// their staking rewards over the long run. Validators can opt out of it by disabling reward restaking on their own StakeEntry.
+	// 2. It simplifies the validator commission distribution code by re-using the same code path for distributing staking
+	// rewards. By requiring the validator to already have a StakeEntry for themselves if they want to restake their commissions,
+	// this approach allows us to avoid manually creating new StakeEntries for the validator specifically for restaking commissions.
+	//
+	// TODO: The downside of the above is that it couples the restaking behavior for validator commissions and the validator's own
+	// staking reward. This is fine though, because if the validator wants to restake their own rewards but not their commissions, then
+	// they can stake to themselves using a separate wallet and only enable reward restaking for that StakeEntry.
+	//
+	// If the above isn't desired the behavior, then we can alternatively always pay out validator's commission directly to their wallet.
```

Comment:
one of the other long term benefits of how this code is set up is that we can track how much a validator got from commissions from a specific staker (if we simply pass in the staker's PKID when we call this function) when we do some utxo operation refactoring work.